cpmpy/solvers/pysdd.py

```diff
@@ -238,42 +226,108 @@ def __add__(self, cpm_expr_orig):
+                                                self._pysdd_expr(cpm_con))
+
+        return self
 
-        elif isinstance(cpm_expr, _BoolVarImpl):
+    def _pysdd_expr(self, cpm_con):
+        """
+            PySDD supports nested expressions: each expression
+            (variable or subexpression) is a node...
+            so we recursively translate our expressions to theirs.
+
+            input: Expression or const
+            output: pysdd Node
+        """
+        if isinstance(cpm_con, _BoolVarImpl):
             # base case, just var or ~var
-            self.pysdd_root &= self.solver_var(cpm_expr)
+            return self.solver_var(cpm_con)
+
+        elif is_bool(cpm_con) or isinstance(cpm_con, BoolVal):
+            # base case: Boolean value
+            if cpm_con:
+                return self.pysdd_manager.true()
+            else:
+                return self.pysdd_manager.false()
+        
+        elif not isinstance(cpm_con, Expression):
+            # a number or so 
+            raise NotImplementedError(f"CPM_pysdd: Non supported object {cpm_con}")
+
+        elif cpm_con.name == 'and':
+            # conjoin the nodes corresponding to the args
+            # also here the order might matter on runtime efficiency...
+            return reduce(self.pysdd_manager.conjoin, [self._pysdd_expr(a) for a in cpm_con.args])
+
+        elif cpm_con.name == 'or':
+            # disjoin the nodes corresponding to the args
+            # also here the order might matter on runtime efficiency...
+            return reduce(self.pysdd_manager.disjoin, [self._pysdd_expr(a) for a in cpm_con.args])
+
+        elif cpm_con.name == 'not':
+            return self.pysdd_manager.negate(self._pysdd_expr(cpm_con.args[0]))
+
+        elif cpm_con.name == '->':
+            a0 = self._pysdd_expr(cpm_con.args[0])
+            a1 = self._pysdd_expr(cpm_con.args[1])
+            # ~a0 | a1
+            return self.pysdd_manager.disjoin(self.pysdd_manager.negate(a0), a1)
+
+        elif cpm_con.name == '==':
+            a0 = self._pysdd_expr(cpm_con.args[0])
+            a1 = self._pysdd_expr(cpm_con.args[1])
+            # (~a0 | a1) & (~a1 | a0)
+            return self.pysdd_manager.conjoin(
+                        self.pysdd_manager.disjoin(self.pysdd_manager.negate(a0), a1),
+                        self.pysdd_manager.disjoin(self.pysdd_manager.negate(a1), a0),
+                   )
+
+        elif cpm_con.name == '!=':
+            # ~(a0 == a1)
+            equiv = self._pysdd_expr(cpm_con.args[0] == cpm_con.args[1]) 
+            return self.pysdd_manager.negate(equiv)
+            
+        elif hasattr(cpm_con, 'decompose'):
+            # for all global constraints: attempt to convert the decomposition
+            all_cons = cpm_con.decompose()  # all of these must be true, conjoin
+            return reduce(self.pysdd_manager.conjoin, [self._pysdd_expr(a) for a in all_cons])
```

Comment:
Will this work with the new decompose globals that gives global and local constraints?